extensions/sftp/processors/ListSFTP.cpp

```diff
@@ -0,0 +1,1320 @@
+  }
+  if (!context->getProperty(StrictHostKeyChecking.getName(), value)) {
+    logger_->log_error("Strict Host Key Checking attribute is missing or invalid");
+  } else {
+    utils::StringUtils::StringToBool(value, strict_host_checking_);
+  }
+  context->getProperty(HostKeyFile.getName(), host_key_file_);
+  if (!context->getProperty(ConnectionTimeout.getName(), value)) {
+    logger_->log_error("Connection Timeout attribute is missing or invalid");
+  } else {
+    core::TimeUnit unit;
+    if (!core::Property::StringToTime(value, connection_timeout_, unit) || !core::Property::ConvertTimeUnitToMS(connection_timeout_, unit, connection_timeout_)) {
+      logger_->log_error("Connection Timeout attribute is invalid");
+    }
+  }
+  if (!context->getProperty(DataTimeout.getName(), value)) {
+    logger_->log_error("Data Timeout attribute is missing or invalid");
+  } else {
+    core::TimeUnit unit;
+    if (!core::Property::StringToTime(value, data_timeout_, unit) || !core::Property::ConvertTimeUnitToMS(data_timeout_, unit, data_timeout_)) {
+      logger_->log_error("Data Timeout attribute is invalid");
+    }
+  }
+  if (!context->getProperty(SendKeepaliveOnTimeout.getName(), value)) {
+    logger_->log_error("Send Keep Alive On Timeout attribute is missing or invalid");
+  } else {
+    utils::StringUtils::StringToBool(value, use_keepalive_on_timeout_);
+  }
+  context->getProperty(TargetSystemTimestampPrecision.getName(), target_system_timestamp_precision_);
+  context->getProperty(ProxyType.getName(), proxy_type_);
+  context->getProperty(EntityTrackingInitialListingTarget.getName(), entity_tracking_initial_listing_target_);
+  if (!context->getProperty(MinimumFileAge.getName(), value)) {
+    logger_->log_error("Minimum File Age attribute is missing or invalid");
+  } else {
+    core::TimeUnit unit;
+    if (!core::Property::StringToTime(value, minimum_file_age_, unit) || !core::Property::ConvertTimeUnitToMS(minimum_file_age_, unit, minimum_file_age_)) {
+      logger_->log_error("Minimum File Age attribute is invalid");
+    }
+  }
+  if (context->getProperty(MaximumFileAge.getName(), value)) {
+    core::TimeUnit unit;
+    if (!core::Property::StringToTime(value, maximum_file_age_, unit) || !core::Property::ConvertTimeUnitToMS(maximum_file_age_, unit, maximum_file_age_)) {
+      logger_->log_error("Maximum File Age attribute is invalid");
+    }
+  }
+  if (!context->getProperty(MinimumFileSize.getName(), minimum_file_size_)) {
+    logger_->log_error("Minimum File Size attribute is invalid");
+  }
+  if (context->getProperty(MaximumFileSize.getName(), value)) {
+    if (!core::DataSizeValue::StringToInt(value, maximum_file_size_)) {
+      logger_->log_error("Maximum File Size attribute is invalid");
+    }
+  }
+  context->getProperty(StateFile.getName(), value);
+  if (listing_strategy_ == LISTING_STRATEGY_TRACKING_TIMESTAMPS) {
+    std::stringstream ss;
+    ss << value << "." << getUUIDStr() << ".TrackingTimestamps";
```

Comment:
I think that the work in https://issues.apache.org/jira/browse/MINIFICPP-550 would have been useful for cases like this. Obviously this ticket will be merged well before that is complete but this confirms in my mind that having state stored within a generalized key/value store backed by rocksdb ( or a volatile store as an option ) would be useful.

Reply:
I agree that such a backend would be useful, and I tried to write the state storage for this processor with two things in my mind: that it should be logically as similar to NiFi as possible, and to make it easily adaptable to a future generic key-value store.